Description: Run nice() before giving up root with setuid()

The daemon code currently tries to nice(-1) just after having given up root
privileges, which fails. This patch moves the nice(-1) call to just before
the code that gives up the required privileges.

Author: Chris Boot <bootc@bootc.net>
Forwarded: http://git.netfilter.org/ulogd2/commit/?id=401c446ce73dfee19afb668a2761f335bcb91877
Reviewed-by: Eric Leblond <eric@regit.org>
Last-Update: 2013-05-21

--- ulogd2-2.0.2.orig/src/ulogd.c
+++ ulogd2-2.0.2/src/ulogd.c
@@ -1235,6 +1235,13 @@ int main(int argc, char* argv[])
 		warn_and_exit(daemonize);
 	}
 
+	errno = 0;
+	if (nice(-1) == -1) {
+		if (errno != 0)
+			ulogd_log(ULOGD_ERROR, "Could not nice process: %s\n",
+				  strerror(errno));
+	}
+
 	if (change_uid) {
 		ulogd_log(ULOGD_NOTICE, "Changing UID / GID\n");
 		if (setgid(gid)) {
@@ -1261,13 +1268,6 @@ int main(int argc, char* argv[])
 		}
 	}
 
-	errno = 0;
-	if (nice(-1) == -1) {
-		if (errno != 0)
-			ulogd_log(ULOGD_ERROR, "Could not nice process: %s\n",
-				  strerror(errno));
-	}
-
 
 	if (daemonize){
 		if (fork()) {
